File: Adidas/adidas-fliteTrakr/java/src/main/java/com/nestis/adidas/fliteTrakr/model/strategy/BelowPriceStrategy.java
package com.nestis.adidas.fliteTrakr.model.strategy;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Defines a Strategy that complies with the maximum price for a route.
 * #9: Find all conenctions from NUE to LHR below 170Euros!
 * @author nestis
 *
 */
public class BelowPriceStrategy extends PriceFlexibleRouteTemplate {
	
	/**
	 * Class constructor.
	 * @param origin Origin
	 * @param dest Destination
	 * @param maxPrice Maximum price for the route.
	 */
	public BelowPriceStrategy(String origin, String dest, Float maxPrice) {
		super(origin, dest, maxPrice);
	}

	@Override
	public String getResult(List<Route> routeFound) {
		
		if (routeFound.isEmpty()) {
			return "No such connection found!";
		} else {
			List<String> routes = routeFound.stream()
				.sorted((a,b) -> a.total.intValue() - b.total.intValue())
				.map(r -> String.join("-", r.route) + "-" + r.total)
				.collect(Collectors.toList());
			return String.join(" ", routes);
		}
	}

}

File: Adidas/adidas-fliteTrakr/js/src/model/strategy/belowPriceStrategy.js
const PriceRouteTemplate = require('./priceRouteTemplate');

/**
 * BelowPriceStrategy function. It will return all the routes found between two cities
 * with a lower price than the given one.
 * @param {*} origin Origin airport.
 * @param {*} dest Destination airport.
 * @param {*} price Maximum price allowed.
 */
function BelowPriceStrategy(origin, dest, price) {
    PriceRouteTemplate.call(this, origin, dest, price);
}

BelowPriceStrategy.prototype = Object.create(PriceRouteTemplate.prototype);
BelowPriceStrategy.prototype.constructor = BelowPriceStrategy;

BelowPriceStrategy.prototype.getInfo = function() {
	if (this.routeFound.length > 0) {
		const routes = [];
		this.routeFound.forEach(r => {
			routes.push(r.path.join('-') + '-' + r.total);
		});
		return routes.join(' ');
	} else {
		return 'No such route found!';
	}
}

module.exports = BelowPriceStrategy;
